From 24a4de6738eb3e00036b00e83077f4116585bb6a Mon Sep 17 00:00:00 2001
From: Christian Ehrhardt <christian.ehrhardt@codasip.com>
Date: Fri, 6 Sep 2024 10:54:34 +0200
Subject: Fix pointer handling in busybox's vi

The upstream code uses realloc to increase an existing memory
region and then adds the pointer difference (bias) between the
old and the new pointer the old pointer to obtain the pointer
in the reallocated region. With CHERI this fails because the
resulting pointer is out of bounds with respect to the origin
allocation. The situation is complicated by the fact that the
calculated bias is returned from the function that calculates
it and callers use it to fixup their pointers.

Fix fix this with minimal changes to uptream keep the pointer
calculations done by upstream but post-process the resulting
pointers. Post processing calculates the offset of the pointer
within the reallocated region and adds this offset to the base
of the region thus using pointer bounds of the new regions
without any change to the address.

Upstream-Status: Pending
---
 editors/vi.c | 34 +++++++++++++++++++++-------------
 1 file changed, 21 insertions(+), 13 deletions(-)

diff --git a/editors/vi.c b/editors/vi.c
index f8fab7028..8e5579600 100644
--- a/editors/vi.c
+++ b/editors/vi.c
@@ -1379,6 +1379,8 @@ static char *swap_context(char *p) // goto new context for '' command make this
 static void undo_push(char *, unsigned, unsigned char);
 #endif
 
+#define F(x) do { (x) = text + ((x) - text); } while (0)
+
 // open a hole in text[]
 // might reallocate text[]! use p += text_hole_make(p, ...),
 // and be careful to not use pointers into potentially freed text[]!
@@ -1394,19 +1396,21 @@ static uintptr_t text_hole_make(char *p, int size)	// at "p", make a 'size' byte
 		text_size += end - (text + text_size) + 10240;
 		new_text = xrealloc(text, text_size);
 		bias = (new_text - text);
-		screenbegin += bias;
-		dot         += bias;
-		end         += bias;
-		p           += bias;
+		text = new_text;
+		screenbegin += bias; F(screenbegin);
+		dot         += bias; F(dot);
+		end         += bias; F(end);
+		p           += bias; F(p);
 #if ENABLE_FEATURE_VI_YANKMARK
 		{
 			int i;
 			for (i = 0; i < ARRAY_SIZE(mark); i++)
-				if (mark[i])
+				if (mark[i]) {
 					mark[i] += bias;
+					F(mark[i]);
+				}
 		}
 #endif
-		text = new_text;
 	}
 	memmove(p + size, p, end - size - p);
 	memset(p, ' ', size);	// clear new hole
@@ -1887,7 +1891,7 @@ static int file_insert(const char *fn, char *p, int initial)
 		goto fi;
 	}
 	size = (statbuf.st_size < INT_MAX ? (int)statbuf.st_size : INT_MAX);
-	p += text_hole_make(p, size);
+	p += text_hole_make(p, size); F(p);
 	cnt = full_read(fd, p, size);
 	if (cnt < 0) {
 		status_line_bold_errno(fn);
@@ -1971,7 +1975,7 @@ static uintptr_t stupid_insert(char *p, char c) // stupidly insert the char c at
 {
 	uintptr_t bias;
 	bias = text_hole_make(p, 1);
-	p += bias;
+	p += bias; F(p);
 	*p = c;
 	return bias;
 }
@@ -1983,6 +1987,7 @@ static char *char_insert(char *p, char c, int undo) // insert the char c at 'p'
 {
 	if (c == 22) {		// Is this an ctrl-V?
 		p += stupid_insert(p, '^');	// use ^ to indicate literal next
+		F(p);
 		refresh(FALSE);	// show the ^
 		c = get_one_char();
 		*p = c;
@@ -2020,6 +2025,7 @@ static char *char_insert(char *p, char c, int undo) // insert the char c at 'p'
 		modified_count++;
 #endif
 		p += 1 + stupid_insert(p, c);	// insert the char
+		F(p);
 #if ENABLE_FEATURE_VI_SETOPTS
 		if (showmatch && strchr(")]}", c) != NULL) {
 			showmatching(p - 1);
@@ -2032,8 +2038,8 @@ static char *char_insert(char *p, char c, int undo) // insert the char c at 'p'
 			if (len) {
 				uintptr_t bias;
 				bias = text_hole_make(p, len);
-				p += bias;
-				q += bias;
+				p += bias; F(p);
+				q += bias; F(q);
 #if ENABLE_FEATURE_VI_UNDO
 				undo_push_insert(p, len, undo);
 #endif
@@ -2095,7 +2101,7 @@ static uintptr_t string_insert(char *p, const char *s, int undo) // insert the s
 	undo_push_insert(p, i, undo);
 #endif
 	bias = text_hole_make(p, i);
-	p += bias;
+	p += bias; F(p);
 	memcpy(p, s, i);
 #if ENABLE_FEATURE_VI_YANKMARK
 	{
@@ -2796,9 +2802,10 @@ static void colon(char *buf)
 #  endif
 				// insert the "replace" patern
 				bias = string_insert(found, R, ALLOW_UNDO_CHAIN);
-				found += bias;
-				ls += bias;
+				found += bias; F(found);
+				ls += bias; F(ls);
 				//q += bias; - recalculated anyway
+				//F(q);
 				// check for "global"  :s/foo/bar/g
 				if (gflag == 'g') {
 					if ((found + len_R) < end_line(ls)) {
@@ -3341,6 +3348,7 @@ static void do_cmd(int c)
 			q = end_line(dot);
 			p = text_hole_delete(p, q, ALLOW_UNDO);	// delete cur line
 			p += string_insert(p, reg[Ureg], ALLOW_UNDO_CHAIN);	// insert orig line
+			F(p);
 			dot = p;
 			dot_skip_over_ws();
 		}
-- 
2.43.5

